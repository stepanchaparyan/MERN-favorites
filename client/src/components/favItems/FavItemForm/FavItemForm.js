import React, { useContext, useState, useEffect, useRef } from 'react';
import { useIntl } from 'react-intl';
import FavItemContext from '../../../context/favItemContext/favItemContext';
import {
  Container,
  Title,
  Form,
  Input,
  Errors,
  ErrorButton,
  Select,
  Option,
  DefaultOption
} from './FavItemFormStyled';
import localization from './localization';

const FavItemForm = () => {
  const context = useContext(FavItemContext);
  const {
    addFavItem,
    editFavItem,
    clearEdit,
    update_FavItem,
    toggle_Form,
    toggleForm,
    error,
    clearErrors
  } = context;

  const { formatMessage } = useIntl();
  const container = useRef();

  useEffect(() => {
    if (editFavItem !== null) {
      setFavItem(editFavItem);
    } else {
      setFavItem({
        author: 'Unknown',
        title: '',
        category: 'Other',
        description: 'Description'
      });
    }
    document.addEventListener('click', handleClick);
    return () => {
      document.removeEventListener('click', handleClick);
    };
  }, [editFavItem, context]);

  const [favItem, setFavItem] = useState({
    author: '',
    title: '',
    category: '',
    description: ''
  });
  const { author, title, category, description } = favItem;
  const onchange = e => {
    setFavItem({
      ...favItem,
      [e.target.name]: e.target.value
    });
  };
  const onsubmit = e => {
    e.preventDefault();
    if (editFavItem === null) {
      addFavItem(favItem);
    } else {
      update_FavItem(favItem);
    }
  };
  const cancelEdit = () => {
    clearEdit();
    toggle_Form(!toggleForm);
  };

  const handleClick = e => {
    if (!container.current.contains(e.target)) {
      toggle_Form(false);
      return;
    }
  };

  const removeErrors = () => {
    clearErrors();
  };

  return (
    <Container ref={container}>
      <Title>
        {editFavItem !== null
          ? formatMessage(localization.editFavItem)
          : formatMessage(localization.addNewFavItem)}
      </Title>
      <Form onSubmit={onsubmit}>
        <Input
          type="text"
          placeholder={formatMessage(localization.author)}
          name="author"
          value={author}
          onChange={onchange}
        />
        <Input
          type="text"
          placeholder={formatMessage(localization.title)}
          name="title"
          value={title}
          onChange={onchange}
          required
        />
        <Select value={category} name="category" onChange={onchange}>
          <DefaultOption value="Other">{formatMessage(localization.selectCategory)}</DefaultOption>
          <Option value="Film">{formatMessage(localization.films)}</Option>
          <Option value="Music">{formatMessage(localization.music)}</Option>
          <Option value="Books">{formatMessage(localization.books)}</Option>
          <Option value="Other">{formatMessage(localization.other)}</Option>
        </Select>
        <Input
          type="text"
          placeholder={formatMessage(localization.description)}
          name="description"
          value={description}
          onChange={onchange}
        />
        <Input
          type="submit"
          value={
            editFavItem !== null
              ? formatMessage(localization.update)
              : formatMessage(localization.add)
          }
        />
        {editFavItem && (
          <Input onClick={cancelEdit} type="button" value={formatMessage(localization.cancel)} />
        )}
      </Form>
      {error && (
        <Errors>
          {!Array.isArray(error) ? (
            <ErrorButton type="button" onClick={removeErrors}>
              {error}
            </ErrorButton>
          ) : (
            error.map(err => (
              <ErrorButton key={err.msg} type="button" onClick={removeErrors}>
                {err.msg}
              </ErrorButton>
            ))
          )}
        </Errors>
      )}
    </Container>
  );
};

export default FavItemForm;
