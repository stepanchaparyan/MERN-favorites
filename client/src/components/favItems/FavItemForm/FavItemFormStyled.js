import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  color: ${props => props.theme.cadetblue};
  z-index: 1;
  position: absolute;
  background-color: gray;
  left: 50%;
  top: 42%;
  transform: translate(-50%, -50%);
  border-radius: 4px;
  min-width: 40%;
  min-height: 40%;
`;

export const Title = styled.div`
  display: flex;
  justify-content: center;
  padding: 32px;
  font-weight: 600;
  font-size: 30px;
  color: black;
`;

export const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const Input = styled.input`
  border: 0px;
  width: 200px;
  padding: 8px 10px;
  margin: 4px;
  outline: none;
  border-radius: 4px;

  &[type='submit'] {
    width: 20%;
    border: 0;
    background-color: ${props => props.theme.cadetblue};
    font-weight: 600;
    font-size: 16px;
    margin-top: 20px;
    cursor: pointer;
    margin-bottom: 10px;
  }

  &[type='button'] {
    width: 30%;
    border: 0;
    background-color: gray;
    font-weight: 600;
    font-size: 16px;
    cursor: pointer;
    margin-bottom: 10px;
  }
`;

export const Errors = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 0 10px 10px 10px;
`;

export const ErrorButton = styled.button`
  color: red;
  background-color: transparent;
  border: none;
  outline: none;
  cursor: pointer;
`;

export const Select = styled.select`
  width: 220px;
  background: white;
  height: 32px;
  outline: none;
  border: none;
  border-radius: 4px;
  padding-left: 5px;
  margin: 4px;
`;

export const Option = styled.option`
  color: black;
  height: 47px;
  padding: 20px;
  font-size: 18px;
  color: ${props => props.theme.cadetblue};
`;

export const DefaultOption = styled.option`
  display: none;
`;
