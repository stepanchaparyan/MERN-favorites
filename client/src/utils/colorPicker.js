export const colorPicker = category => {
  let color;
  switch (category) {
    case 'Film':
      color = 'red';
      break;
    case 'Music':
      color = 'blue';
      break;
    case 'Books':
      color = 'limeGreen';
      break;
    case 'Other':
      color = 'orange';
      break;
    default:
      color = 'black';
  }
  return color;
};
