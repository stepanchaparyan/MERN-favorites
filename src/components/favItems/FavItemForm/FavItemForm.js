import React, { useContext, useState, useEffect, useRef } from 'react';
import FavItemContext from '../../../context/favItemContext/favItemContext';
import {
  Container,
  Title,
  Form,
  Input,
  Errors,
  ErrorButton,
  Select,
  Option,
  DefaultOption
} from './FavItemFormStyled';

const FavItemForm = () => {
  const context = useContext(FavItemContext);
  const {
    addFavItem,
    editFavItem,
    clearEdit,
    update_FavItem,
    toggle_Form,
    toggleForm,
    error,
    clearErrors
  } = context;

  const container = useRef();

  useEffect(() => {
    if (editFavItem !== null) {
      setFavItem(editFavItem);
    } else {
      setFavItem({
        author: 'Unknown',
        title: '',
        category: 'Other',
        description: 'Description'
      });
    }
    document.addEventListener('click', handleClick);
    return () => {
      document.removeEventListener('click', handleClick);
    };
  }, [editFavItem, context]);

  const [favItem, setFavItem] = useState({
    author: '',
    title: '',
    category: '',
    description: ''
  });
  const { author, title, category, description } = favItem;
  const onchange = e => {
    setFavItem({
      ...favItem,
      [e.target.name]: e.target.value
    });
  };
  const onsubmit = e => {
    e.preventDefault();
    if (editFavItem === null) {
      addFavItem(favItem);
    } else {
      update_FavItem(favItem);
      // clearEdit();
      // toggle_Form(!toggleForm);
    }
    setFavItem({
      author: '',
      title: '',
      category: '',
      description: ''
    });
  };
  const cancelEdit = () => {
    clearEdit();
    toggle_Form(!toggleForm);
  };

  const handleClick = e => {
    if (!container.current.contains(e.target)) {
      toggle_Form(false);
      return;
    }
  };

  const removeErrors = () => {
    // toggle_Form(true);
    clearErrors();
  };

  return (
    <Container ref={container}>
      <Title>{editFavItem !== null ? 'Edit FavItem' : 'Add new favorite item'}</Title>
      <Form onSubmit={onsubmit}>
        <Input type="text" placeholder="Author" name="author" value={author} onChange={onchange} />
        <Input
          type="text"
          placeholder="Title"
          name="title"
          value={title}
          onChange={onchange}
          required
        />
        <Select value={category} name="category" onChange={onchange}>
          <DefaultOption value="Other">Select category</DefaultOption>
          <Option value="Film">Film</Option>
          <Option value="Music">Music</Option>
          <Option value="Books">Books</Option>
          <Option value="Other">Other</Option>
        </Select>
        <Input
          type="text"
          placeholder="Description"
          name="description"
          value={description}
          onChange={onchange}
        />
        <Input type="submit" value={editFavItem !== null ? 'Update' : 'Add'} />
        {editFavItem && <Input onClick={cancelEdit} type="button" value="Cancel" />}
      </Form>
      {error && (
        <Errors>
          {!Array.isArray(error) ? (
            <ErrorButton type="button" onClick={removeErrors}>
              {error}
            </ErrorButton>
          ) : (
            error.map(err => (
              <ErrorButton key={err.msg} type="button" onClick={removeErrors}>
                {err.msg}
              </ErrorButton>
            ))
          )}
        </Errors>
      )}
    </Container>
  );
};

export default FavItemForm;
