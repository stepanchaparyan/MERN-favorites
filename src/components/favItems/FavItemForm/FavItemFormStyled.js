import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  color: cadetblue;
  z-index: 1;
  position: absolute;
  background-color: darkmagenta;
  left: 50%;
  top: 40%;
  transform: translate(-50%, -50%);
  border-radius: 4px;
`;

export const Title = styled.div`
  display: flex;
  justify-content: center;
  padding: 32px;
  font-weight: 600;
  font-size: 30px;
`;

export const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const Input = styled.input`
  border: 0px;
  border-bottom: 1px solid cadetblue;
  width: 30%;
  padding: 7px 10px;
  margin: 2px;
  outline: none;

  &[type='submit'] {
    width: 20%;
    border: 0;
    background-color: cadetblue;
    font-weight: 600;
    font-size: 16px;
    margin-top: 20px;
    border-radius: 3%;
    cursor: pointer;
  }
`;

export const Errors = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 10px;
`;

export const ErrorButton = styled.button`
  color: red;
  background-color: transparent;
  border: none;
  outline: none;
  cursor: pointer;
`;
