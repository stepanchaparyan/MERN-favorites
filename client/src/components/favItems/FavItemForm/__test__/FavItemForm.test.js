import React, { useContext } from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import { renderSnapshot } from '../../../../utils/tests';
import FavItemForm from '../FavItemForm.js';
import { Form, CancelButton, Input, ErrorButton, Select } from '../FavItemFormStyled';

jest.mock('react-intl', () => ({
  useIntl: jest.fn(() => ({ formatMessage: m => m.defaultMessage })),
  defineMessages: i => i
}));

jest.mock('react', () => {
  const ActualReact = jest.requireActual('react');
  return {
    ...ActualReact,
    useContext: jest.fn().mockImplementation(() => ({
      onSubmit: jest.fn(),
      update_FavItem: jest.fn(),
      error: 'a',
      clearErrors: jest.fn()
    }))
  };
});

describe('FavItemForm component test with Enzyme', () => {
  test('renders correct snapshot', () => {
    renderSnapshot(FavItemForm);
  });

  test('render submit on Form component', () => {
    const component = shallow(<FavItemForm />);
    const form = component.find(Form);

    const preventDefault = jest.fn();
    form.simulate('submit', { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
  });

  test('render submit on Form component when editFavitem is true', () => {
    useContext.mockImplementationOnce(
      jest.fn(() => ({
        editFavItem: {},
        clearEdit: jest.fn(),
        toggle_Form: jest.fn(),
        update_FavItem: jest.fn()
      }))
    );

    const component = shallow(<FavItemForm />);

    const cancelButton = component.find(CancelButton);
    cancelButton.simulate('click');

    const form = component.find(Form);
    const preventDefault = jest.fn();
    form.simulate('submit', { preventDefault });
    expect(preventDefault).toHaveBeenCalled();

    expect(component).toMatchSnapshot();
  });
});

test('render change on Input component', () => {
  useContext.mockImplementationOnce(
    jest.fn(() => ({
      editFavItem: {},
      clearEdit: jest.fn(),
      toggle_Form: jest.fn(),
      update_FavItem: jest.fn(),
      clearErrors: jest.fn()
    }))
  );

  const component = shallow(<FavItemForm />);

  const inputs = component.find(Input);
  inputs.forEach(input => {
    input.simulate('change', { target: { value: 'target' } });
  });

  const select = component.find(Select);
  select.simulate('change', { target: { value: 'target' } });

  expect(component).toMatchSnapshot();
});

test('render click on ErrorButton when error is not array', () => {
  const component = shallow(<FavItemForm />);

  const errorButtons = component.find(ErrorButton);
  errorButtons.forEach(errorButton => {
    errorButton.simulate('click');
  });
  expect(component).toMatchSnapshot();
});

test('render click on ErrorButton when error is array', () => {
  useContext.mockImplementationOnce(
    jest.fn(() => ({
      editFavItem: {},
      clearEdit: jest.fn(),
      toggle_Form: jest.fn(),
      update_FavItem: jest.fn(),
      clearErrors: jest.fn(),
      error: ['a', 'b']
    }))
  );

  const component = shallow(<FavItemForm />);

  const errorButtons = component.find(ErrorButton);
  errorButtons.forEach(errorButton => {
    errorButton.simulate('click');
  });
  expect(component).toMatchSnapshot();
});
