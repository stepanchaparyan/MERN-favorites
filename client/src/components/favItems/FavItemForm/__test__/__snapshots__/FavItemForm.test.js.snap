// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavItemForm component test with Enzyme render submit on Form component when editFavitem is true 1`] = `
<styled.div>
  <styled.div>
    Edit favorite item
  </styled.div>
  <styled.form
    onSubmit={[Function]}
  >
    <styled.input
      name="author"
      onChange={[Function]}
      placeholder="Author"
      type="text"
      value=""
    />
    <styled.input
      name="title"
      onChange={[Function]}
      placeholder="Title"
      required={true}
      type="text"
      value=""
    />
    <styled.select
      name="category"
      onChange={[Function]}
      value=""
    >
      <styled.option
        value="Other"
      >
        Select Category
      </styled.option>
      <styled.option
        value="Film"
      >
        Films
      </styled.option>
      <styled.option
        value="Music"
      >
        Music
      </styled.option>
      <styled.option
        value="Books"
      >
        Books
      </styled.option>
      <styled.option
        value="Other"
      >
        Other
      </styled.option>
    </styled.select>
    <styled.input
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <styled.button>
      Update
    </styled.button>
    <styled.button
      onClick={[Function]}
    >
      Cancel
    </styled.button>
  </styled.form>
</styled.div>
`;

exports[`FavItemForm component test with Enzyme renders correct snapshot 1`] = `
<styled.div>
  <styled.div>
    Edit favorite item
  </styled.div>
  <styled.form
    onSubmit={[Function]}
  >
    <styled.input
      name="author"
      onChange={[Function]}
      placeholder="Author"
      type="text"
      value=""
    />
    <styled.input
      name="title"
      onChange={[Function]}
      placeholder="Title"
      required={true}
      type="text"
      value=""
    />
    <styled.select
      name="category"
      onChange={[Function]}
      value=""
    >
      <styled.option
        value="Other"
      >
        Select Category
      </styled.option>
      <styled.option
        value="Film"
      >
        Films
      </styled.option>
      <styled.option
        value="Music"
      >
        Music
      </styled.option>
      <styled.option
        value="Books"
      >
        Books
      </styled.option>
      <styled.option
        value="Other"
      >
        Other
      </styled.option>
    </styled.select>
    <styled.input
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <styled.button>
      Update
    </styled.button>
  </styled.form>
  <styled.div>
    <styled.button
      onClick={[Function]}
      type="button"
    >
      a
    </styled.button>
  </styled.div>
</styled.div>
`;

exports[`render change on Input component 1`] = `
<styled.div>
  <styled.div>
    Edit favorite item
  </styled.div>
  <styled.form
    onSubmit={[Function]}
  >
    <styled.input
      name="author"
      onChange={[Function]}
      placeholder="Author"
      type="text"
      value=""
    />
    <styled.input
      name="title"
      onChange={[Function]}
      placeholder="Title"
      required={true}
      type="text"
      value=""
    />
    <styled.select
      name="category"
      onChange={[Function]}
      value=""
    >
      <styled.option
        value="Other"
      >
        Select Category
      </styled.option>
      <styled.option
        value="Film"
      >
        Films
      </styled.option>
      <styled.option
        value="Music"
      >
        Music
      </styled.option>
      <styled.option
        value="Books"
      >
        Books
      </styled.option>
      <styled.option
        value="Other"
      >
        Other
      </styled.option>
    </styled.select>
    <styled.input
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <styled.button>
      Update
    </styled.button>
  </styled.form>
  <styled.div>
    <styled.button
      onClick={[Function]}
      type="button"
    >
      a
    </styled.button>
  </styled.div>
</styled.div>
`;

exports[`render click on ErrorButton when error is array 1`] = `
<styled.div>
  <styled.div>
    Edit favorite item
  </styled.div>
  <styled.form
    onSubmit={[Function]}
  >
    <styled.input
      name="author"
      onChange={[Function]}
      placeholder="Author"
      type="text"
      value=""
    />
    <styled.input
      name="title"
      onChange={[Function]}
      placeholder="Title"
      required={true}
      type="text"
      value=""
    />
    <styled.select
      name="category"
      onChange={[Function]}
      value=""
    >
      <styled.option
        value="Other"
      >
        Select Category
      </styled.option>
      <styled.option
        value="Film"
      >
        Films
      </styled.option>
      <styled.option
        value="Music"
      >
        Music
      </styled.option>
      <styled.option
        value="Books"
      >
        Books
      </styled.option>
      <styled.option
        value="Other"
      >
        Other
      </styled.option>
    </styled.select>
    <styled.input
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <styled.button>
      Update
    </styled.button>
    <styled.button
      onClick={[Function]}
    >
      Cancel
    </styled.button>
  </styled.form>
  <styled.div>
    <styled.button
      onClick={[Function]}
      type="button"
    />
    <styled.button
      onClick={[Function]}
      type="button"
    />
  </styled.div>
</styled.div>
`;

exports[`render click on ErrorButton when error is not array 1`] = `
<styled.div>
  <styled.div>
    Edit favorite item
  </styled.div>
  <styled.form
    onSubmit={[Function]}
  >
    <styled.input
      name="author"
      onChange={[Function]}
      placeholder="Author"
      type="text"
      value=""
    />
    <styled.input
      name="title"
      onChange={[Function]}
      placeholder="Title"
      required={true}
      type="text"
      value=""
    />
    <styled.select
      name="category"
      onChange={[Function]}
      value=""
    >
      <styled.option
        value="Other"
      >
        Select Category
      </styled.option>
      <styled.option
        value="Film"
      >
        Films
      </styled.option>
      <styled.option
        value="Music"
      >
        Music
      </styled.option>
      <styled.option
        value="Books"
      >
        Books
      </styled.option>
      <styled.option
        value="Other"
      >
        Other
      </styled.option>
    </styled.select>
    <styled.input
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <styled.button>
      Update
    </styled.button>
  </styled.form>
  <styled.div>
    <styled.button
      onClick={[Function]}
      type="button"
    >
      a
    </styled.button>
  </styled.div>
</styled.div>
`;
